storage/range_tree.go

```diff
@@ -18,31 +18,306 @@
+	}
+	return nil
+}
 
+// SetupRangeTree creates a new RangeTree.  This should only be called as part of
+// store.BootstrapRange.
+func SetupRangeTree(batch engine.Engine, ms *engine.MVCCStats, timestamp proto.Timestamp, startKey proto.Key) error {
 	tree := &proto.RangeTree{
 		RootKey: startKey,
 	}
-
 	node := &proto.RangeTreeNode{
 		Key:   startKey,
 		Black: true,
 	}
-
+	if err := engine.MVCCPutProto(batch, ms, engine.KeyRangeTreeRoot, timestamp, nil, tree); err != nil {
+		return err
+	}
 	if err := engine.MVCCPutProto(batch, ms, engine.RangeTreeNodeKey(startKey), timestamp, nil, node); err != nil {
 		return err
 	}
+	return nil
+}
 
-	if err := engine.MVCCPutProto(batch, ms, engine.KeyRangeTreeRoot, timestamp, nil, tree); err != nil {
-		return err
+// GetRangeTree fetches the RangeTree proto.
+func getRangeTree(db *client.KV) (*rangeTree, error) {
+	tree := &proto.RangeTree{}
+	ok, _, err := db.GetProto(engine.KeyRangeTreeRoot, tree)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, util.Errorf("Could not find the range tree:%s", engine.KeyRangeTreeRoot)
+	}
+	return (*rangeTree)(tree), nil
+}
+
+// getRangeTreeNode return the RangeTree node for the given key.  If the key is
+// nil, an empty RangeTreeNode is returned.
+func getRangeTreeNode(db *client.KV, key *proto.Key) (*rangeTreeNode, error) {
+	if key == nil {
+		return nil, nil
+	}
+	node := &proto.RangeTreeNode{}
+	ok, _, err := db.GetProto(engine.RangeTreeNodeKey(*key), node)
+	if err != nil {
+		return nil, err
 	}
+	if !ok {
+		return nil, util.Errorf("Could not find the range tree node:%s",
+			engine.RangeTreeNodeKey(*key))
+	}
+	return (*rangeTreeNode)(node), nil
+}
 
+// getRootNode returns the RangeTree node for the root of the RangeTree.
+func (t *rangeTree) getRootNode(db *client.KV) (*rangeTreeNode, error) {
+	return getRangeTreeNode(db, &t.RootKey)
+}
+
+// InsertRange adds a new range to the RangeTree.  This should only be called
+// from operations that create new ranges, such as range.splitTrigger.
+func InsertRange(db *client.KV, key proto.Key) error {
+	tree, err := getRangeTree(db)
+	if err != nil {
+		return err
+	}
+	root, err := tree.getRootNode(db)
+	if err != nil {
+		return err
+	}
+	root, err = tree.insert(db, root, key)
+	if err != nil {
+		return err
+	}
+	if !root.Black {
+		// Always set the root back to black.
+		root.Black = true
+		if err = root.set(db); err != nil {
+			return err
+		}
+	}
+	if !tree.RootKey.Equal(root.Key) {
+		// If the root has changed, update the tree to point to it.
+		tree.RootKey = root.Key
+		if err = tree.set(db); err != nil {
+			return err
+		}
+	}
 	return nil
 }
+
+// insert performs the insertion of a new range into the RangeTree.  It will
+// recursivly call insert until it finds the correct location.  It will not
+// overwite an already existing key, but that case should not occur.
+func (t *rangeTree) insert(db *client.KV, node *rangeTreeNode, key proto.Key) (*rangeTreeNode, error) {
```

Comment:
Yeah.  I've found some examples of keeping parent nodes.  This will require the parent key as a parameter at the very least.